nntrainer/tensor/tensor.cpp

```diff
@@ -80,18 +80,80 @@ static auto rng = [] {
   return rng;
 }();
 
-Tensor::Tensor(const TensorDim &d, const float *buf) : Tensor() {
+Tensor::Tensor(const TensorDim &d, const float *buf, bool alloc_now) : Tensor() {
   if (d.getDataLen() != 0) {
+    if (buf != nullptr && !alloc_now)
+      throw std::invalid_argument("alloc_now must be true to create a tensor with a buffer");
+
     dim = d;
     strides = d.computeStrides();
-    data = std::shared_ptr<float>(new float[d.getDataLen()],
-                                  std::default_delete<float[]>());
+
+    if (alloc_now)
+      data = std::shared_ptr<float>(new float[d.getDataLen()],
+          std::default_delete<float[]>());
 
     if (buf != nullptr)
       copy(buf);
   }
 }
 
+/**
+ * @class SrcSharedTensor
+ * @brief Source of the shared tensor
+ */
+class SrcSharedTensor {
+public:
```

Comment:
I wonder.. does the tensor has this SrcSharedTensor always? How about just set the data with data pointer from shared tensor ?

Reply:
> How about just set the data with data pointer from shared tensor ?

That is one approach. However, with this, there need to be two extra variables - offset from srcSharedTensor, and a variable to indicate if this actual data or srcSharedTensor, and creates Tensor object more complicated and heavy.
With SrcSharedTensor, the Tensor class size remains cleaner with smaller increase in object size.